core/internal/api/credentials.go

```diff
@@ -55,3 +72,274 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+		logger:                logger,
+	}, nil
+}
+
+type oauth2CredentialProvider struct {
+	// The URL of the W&B API
+	baseURL string
+
+	// The file path to the JWT
+	identityTokenFilePath string
+
+	// The file path to the access token and its metadata
+	credentialsFilePath string
+
+	// The access token and its metadata
+	token tokenInfo
+
+	tokenMu *sync.RWMutex
+
+	logger *slog.Logger
+}
+
+// ExpiresAt is a custom type representing a time.Time value. It is used to handle
+// expiration times in a specific string format when serializing/deserializing JSON data.
+type ExpiresAt time.Time
+
+const expiresAtLayout = "2006-01-02 15:04:05"
+
+func (e *ExpiresAt) UnmarshalJSON(data []byte) error {
+	var timeString string
+	if err := json.Unmarshal(data, &timeString); err != nil {
+		return err
+	}
+
+	parsedTime, err := time.Parse(expiresAtLayout, timeString)
+	if err != nil {
+		return err
+	}
+
+	*e = ExpiresAt(parsedTime)
+	return nil
+}
+
+func (e ExpiresAt) MarshalJSON() ([]byte, error) {
+	formattedTime := time.Time(e).Format(expiresAtLayout)
+	return json.Marshal(formattedTime)
+}
+
+type tokenInfo struct {
+	// The time at which the token will expire
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *tokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]tokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.shouldRefreshToken() {
+		err := c.loadCredentials()
+		if err != nil {
+			return err
+		}
+	}
+
+	req.Header.Set(
+		"Authorization",
+		"Bearer "+c.token.AccessToken,
+	)
+	return nil
+}
+
+func (c *oauth2CredentialProvider) shouldRefreshToken() bool {
+	c.tokenMu.RLock()
+	defer c.tokenMu.RUnlock()
+
+	return c.token.IsTokenExpiring()
+}
+
+// loadCredentials attempts to load an access token from the credentials file.
+// If the credentials file does not exist, it attempts to create it.
+func (c *oauth2CredentialProvider) loadCredentials() error {
+	c.tokenMu.Lock()
+	defer c.tokenMu.Unlock()
+
+	if !c.token.IsTokenExpiring() {
+		return nil
+	}
+
+	_, err := os.Stat(c.credentialsFilePath)
+	if os.IsNotExist(err) {
+		token, err := c.writeCredentialsFile()
+		if err != nil {
+			return err
+		}
+		c.token = token
+		return nil
+	}
+	if err != nil {
+		c.logger.Warn("failed to access credentials file",
+			"file path", c.credentialsFilePath,
+			"error", err,
+		)
+	}
+
+	token, err := c.loadCredentialsFromFile()
+	if err != nil {
+		return err
+	}
+	c.token = token
+
+	return nil
```

Comment:
The logic here should be simpler and could use better naming. It makes it seem that `loadCredentials` can fail if `writeCredentialsFile` fails, that is, "writing credentials to file" fails. I was about to comment on it, but it's not true, and you can only tell by reading the code.

From what I can tell, it should just be:

1. Try reading unexpired credentials from a file. This fails if the file doesn't exist, the reading operation fails, parsing its contents fails, or the credentials are expired. If it fails, continue on to (2).
2. Load a new access token. This fails if the identity token doesn't exist, or if the network operation fails. If it fails, then there's no access token, so error out.
3. Try to write the token to a file. If this fails, just warn.

```go
creds, ok := tryLoadCredentialsFromFile()
if ok && !creds.IsTokenExpiring() {
	c.token = creds
	return nil
}

creds, err = fetchNewToken()
if err != nil {
	return fmt.Errorf("api: couldn't fetch access token: %v", err)
}

c.token = creds
c.trySaveCredentialsToFile()
return nil
```

You shouldn't have to Stat the file---if the file doesn't exist, reading will fail. `tryLoadCredentialsFromFile` shouldn't do anything other than read a file (it shouldn't write a file, and it shouldn't make network requests).

Reply:
I agree with separating the concerns here. Updated